File: Data Structures & Algorithms/validate-parentheses/submission-4.java
class Solution {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        if ( s.length() % 2 != 0 ) return false;

        for ( int i = 0; i < s.length(); i++) {

            if ( s.charAt(i) == '(') stack.add(')');
            else if ( s.charAt(i) == '[') stack.add(']');
            else if ( s.charAt(i) == '{') stack.add('}');
            else if ( stack.isEmpty() ) return false;
            else if ( stack.pop() != s.charAt(i)) return false;
        }
        if ( !stack.isEmpty() ) return false;
        return true;
    }
}

File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-4.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        operators = {"+", "-", "*", "/"}
        
        for token in tokens:
            if token in operators:
                d2 = stack.pop()
                d1 = stack.pop()
                
                if token == "+":
                    stack.append(d1 + d2)
                elif token == "-":
                    stack.append(d1 - d2)
                elif token == "*":
                    stack.append(d1 * d2)
                elif token == "/":
                    stack.append(int(d1 / d2))
            else:
                stack.append(int(token))
        
        return stack[0]